### DISCUSSION TITLE
Undefined References

### DISCUSSION THREAD

**Discussion by gabrofig** [open-quantum-safe/liboqs#1571] [Q&A]
Hello everyone,
I am trying to apply the digital signature from lioqs algorithm for a school project. 
I followed the instructions to build and believe every step was successful. However, I tried to run in MSCode the _example_sig.c_ to see how it goes and find the following error message:

usr/bin/ld: /tmp/ccbunQVD.o: in function `example_stack':
hello.c:(.text+0x4a): undefined reference to `OQS_randombytes'
/usr/bin/ld: hello.c:(.text+0x63): undefined reference to `OQS_SIG_dilithium_2_keypair'
/usr/bin/ld: hello.c:(.text+0xe1): undefined reference to `OQS_SIG_dilithium_2_sign'
/usr/bin/ld: hello.c:(.text+0x15f): undefined reference to `OQS_SIG_dilithium_2_verify'
 ....

Is it something to do with my build or maybe it is not the way to apply the algorithm properly?
Thank you very much and I appreciate any feedback!
Kind regards

**Comment by ashman-p** [comment]
Hi, 
Just a few quick questions...
Have you verified your build by running any of the tests from build output directory?
Next I would suggest posting your makefile/command line, OQS install location, and
maybe you can post your hello.c.

**Comment by baentsch** [comment]
Please take a look at the questions asked when [opening a bug report](https://github.com/open-quantum-safe/liboqs/issues/new/choose) and let us have that: Without this information we have to guess what's going on in your setup. Also, please consider using a [pre-build docker image](https://hub.docker.com/repository/docker/openquantumsafe/oqs-ossl3/general) giving you a well-known interface (OpenSSLs in this sample case).

**Comment by gabrofig** [comment]
Thank you both for the feedback. 
I followed baentschs suggestion and when trying the Docker Image approach, I found the liboqs-python( which I though would be easier for me). However, I still found issues, this time, with the error `Could not load liboqs shared library`. I checked other issues like [#28](https://github.com/open-quantum-safe/liboqs-python/issues/28) and [#36](https://github.com/open-quantum-safe/liboqs-python/issues/51) but still find the issue even when I make sure my LD_LIBRARY_PATH is set. Even confirmed with _ls $PYTHONPATH_command mentioned in the other post.

Is there anything that I might be doing wrong when setting the LD_LIBRARY_PATH? I attached the command line logs (if it helps).
[logs_build.txt](https://github.com/open-quantum-safe/liboqs/files/12838771/logs_build.txt)

Hope this is enough information and please, let me know if I can provide anything else :)

Thank you very much for the attention!
Kind regards,

**Reply by baentsch** [reply]
> when trying the Docker Image approach, I found the liboqs-python( which I though would be easier for me)

Why do you think this would be easier? I can hardly think of anything simpler for generating (PQ or classic) keys and certs and using them to sign some data than [openssl](https://www.zimuel.it/blog/sign-and-verify-a-file-using-openssl).

**Reply by gabrofig** [reply]
As the project I`m working on is related to build a communication protocol myself implementing post quantum algorithms, I thought using the python approach would allow me more visibility on the code`s steps than with openssl. And since openssl is already working with TLS and SSL, I thought using the python wrapper I could work from scratch on my protocol. Does it make sense or I can achieve the same with openssl?
Thanks and best regards!

**Reply by baentsch** [reply]
Ahh -- you ought to develop some protocol _yourself_ and add PQ to it. Then simply using (Open)SSL of course is way too simple. I've got a ToDo on my own list to put up a ready-to-run `liboqs-python` container for `oqs-demos`. Just moved that up my priority list :) See the file [here](https://github.com/open-quantum-safe/oqs-demos/tree/mb-languagewrappers/language-wrappers/python). Builds as usual with `docker build -t oqs-python .` and you can log in to it with `docker run -it oqs-python sh`. The build may also show where your own build is going astray.

**Reply by gabrofig** [reply]
Many thanks for the insight!
When following exactly the steps:
-Cloning oqs-demos
- running _docker build -t oqs-python ._ in oqs-demos/language-wrappers-python fiile
- _docker run -it oqs-python sh_
- python3 examples/sig.py

I have the warning message _UserWarning: Please install liboqs-python using setup.py_
Or when running test_sig.py or test_kem.py, I get _ModuleNotFoundError: No module named 'nose'_

Anything I might be missing?

I appreciate the attention, thanks a lot.
Kind regards.

**Reply by baentsch** [reply]
> Anything I might be missing?

Yes. You may be running the final command outside of the docker container: Are you? What's your userid when executing `python3 examples/sig.py`? It should be "oqs".

**Reply by gabrofig** [reply]
Dear Michael,
I tried running the command to make sure that I am executing in the docker container but got the same message:
_docker run -it oqs-python sh -c ". python3 liboqs-python/examples/sig.py"_ 


```

/opt/liboqs-python/oqs/oqs.py:68: UserWarning: Please install liboqs-python using setup.py
  warnings.warn("Please install liboqs-python using setup.py")
/opt/liboqs-python/oqs/oqs.py:75: UserWarning: liboqs version 0.9.0 differs from liboqs-python version None
  warnings.warn("liboqs version {} differs from liboqs-python version {}".format(oqs_version(), oqs_python_version()))
liboqs version: 0.9.0
liboqs-python version: None
Enabled signature mechanisms:
['Dilithium2', 'Dilithium3', 'Dilithium5', 'Falcon-512', 'Falcon-1024',
 'SPHINCS+-SHA2-128f-simple', 'SPHINCS+-SHA2-128s-simple',
 'SPHINCS+-SHA2-192f-simple', 'SPHINCS+-SHA2-192s-simple',
 'SPHINCS+-SHA2-256f-simple', 'SPHINCS+-SHA2-256s-simple',
 'SPHINCS+-SHAKE-128f-simple', 'SPHINCS+-SHAKE-128s-simple',
 'SPHINCS+-SHAKE-192f-simple', 'SPHINCS+-SHAKE-192s-simple',
 'SPHINCS+-SHAKE-256f-simple', 'SPHINCS+-SHAKE-256s-simple']

Signature details:
{'claimed_nist_level': 2,
 'is_euf_cma': True,
 'length_public_key': 1312,
 'length_secret_key': 2528,
 'length_signature': 2420,
 'name': 'Dilithium2',
 'version': 'https://github.com/pq-crystals/dilithium/commit/d9c885d3f2e11c05529eeeb7d70d808c972b8409'}

Valid signature? True
```
If this means that my userID is not oqs, as you suggested above, how can I make sure the userID is correctly inserted
Thank you and kind regards,

**Comment by baentsch** [comment]
> If this means that my userID is not oqs, as you suggested above, how can I make sure the userID is correctly inserted

No this does not mean it and you do not need to do anything. The error message is clear: The python binding to liboqs is not in sync with the actual liboqs version. This is just a warning. 

But the logic works (shown in the last line). 

@vsoftco I tend to regret the notion of checking integrity of `liboqs-python` with `liboqs`: This inconsistency happens pretty often now it seems: Either we find a way to automate the establishment of that property or do away with this sanity check and deal with (or don't care about) the possible negative consequences ("surprising" functional differences or crashes when people only use `liboqs-python` on an incompatible pre-installed `liboqs`, e.g., if both are not installed at the same time, e.g., via package managers). 

@dstebila May I suggest a meeting (or put it onto the agenda of a regular meeting) to agree which sub-projects must be in sync with `liboqs` (and/or must be able to deal with pre-installed `liboqs` versions), which ones are optional to be and how we ensure this.
